client/canonical_eventlog.go

```diff
@@ -0,0 +1,266 @@
+		digest, err := payload.GenerateDigest(hashAlgo)
+		if err != nil {
+			return err
+		}
+		digestsMap[hashAlgoMapping[hashAlgo]] = digest
+
+		// TODO: extend the digest to TPM PCR
+	}
+
+	celr, err := createCELR(uint64(len(c.Records)), uint8(pcr), digestsMap, payload)
+	if err != nil {
+		return err
+	}
+
+	c.Records = append(c.Records, celr)
+	return nil
+}
+
+func createRecNumField(recNum uint64) TLV {
+	payload := make([]byte, recNumValueLength)
+	binary.BigEndian.PutUint64(payload, recNum)
+	return TLV{recNumTagValue, recNumValueLength, payload}
+}
+
+// UnmarshalRecNum takes in a TLV with its tag equals to the recnum tag value (0), and
+// return its record number.
+func UnmarshalRecNum(tlv TLV) (uint64, error) {
+	if tlv.Tag != recNumTagValue {
+		return 0, fmt.Errorf("tag of the TLV indicates it is not a recnum field")
+	}
+	return binary.BigEndian.Uint64(tlv.Value), nil
+}
+
+func createPCRField(pcrNum uint8) TLV {
+	return TLV{pcrTagValue, pcrValueLength, []byte{pcrNum}}
+}
+
+// UnmarshalPCR takes in a TLV with its tag equals to the PCR tag value (3), and
+// return its PCR number.
+func UnmarshalPCR(tlv TLV) (pcrNum uint8, err error) {
+	if tlv.Tag != pcrTagValue {
+		return 0, fmt.Errorf("tag of the TLV indicates it is not a PCR field")
+	}
+	return tlv.Value[0], nil
+}
+
+func createDigestField(digestMap map[tpm2.Algorithm][]byte) TLV {
+	var buf bytes.Buffer
+
+	for hashAlgo, hash := range digestMap {
+		singledigest := TLV{uint8(hashAlgo), uint32(len(hash)), hash}
+		_, err := buf.Write(TLVMarshal(singledigest))
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return TLV{digestsTagValue, uint32(buf.Len()), buf.Bytes()}
+}
+
+// UnmarshalDigests takes in a TLV with its tag equals to the digests Tag value (3), and
+// return its digests content in a map, the key is its TPM hash algorithm.
+func UnmarshalDigests(tlv TLV) (digestsMap map[tpm2.Algorithm][]byte, err error) {
+	if tlv.Tag != digestsTagValue {
+		return nil, fmt.Errorf("tag of the TLV indicates it doesn't contain digests")
+	}
+
+	buf := bytes.NewBuffer(tlv.Value)
+	digestsMap = make(map[tpm2.Algorithm][]byte)
+
+	for buf.Len() > 0 {
+		digestTLV, err := TLVUnmarshal(buf)
+		if err == io.EOF {
+			return nil, fmt.Errorf("buffer ends unexpectedly")
+		}
+		digestsMap[tpm2.Algorithm(digestTLV.Tag)] = digestTLV.Value
+	}
+	return digestsMap, nil
+}
+
+func createCELR(recNum uint64, pcr uint8, digestsmap map[tpm2.Algorithm][]byte, event CELREventData) (celr CELR, err error) {
+	recnumField := createRecNumField(recNum)
+	pcrField := createPCRField(pcr)
+
+	digestField := createDigestField(digestsmap)
+	celr = CELR{recnumField, pcrField, digestField, event.GetTLV()}
+
+	return celr, nil
+}
+
+// EncodeCELR encodes the CELR to bytes according to the CEL spec and write them
+// to the byte byffer.
+func (r *CELR) EncodeCELR(buf *bytes.Buffer) error {
+	// encode CEL and write the binary eventlog to the buf
+	recnumField := TLVMarshal(TLV(r.RecNum))
```

Comment:
we should be able to change to `r.Recnum.Marshal()` if we make `TLVMarshal` a receiver method

Reply:
Done